Frameworks/CoreText/DWriteWrapper.mm

```diff
@@ -430,17 +423,26 @@ static void __GetGlyphRunDetails(_CTTypesetter* ts, CFRange range, CGRect frameS
  * @parameter range attributed string range to use.
  * @parameter frameSize size parameters of the frame to fit the text into.
  *
- * @return Unmutable array of _CTLine objects created with the requested parameters.
+ * @return _CTFrame* created using the given parameters
  */
-static NSArray<_CTLine*>* _DWriteGetLines(_CTTypesetter* ts, CFRange range, CGRect frameSize) {
-    NSMutableArray<_CTLine*>* lines = [NSMutableArray array];
+static _CTFrame* _DWriteGetFrame(_CTTypesetter* ts, CFRange range, CGRect frameSize) {
+    _CTFrame* frame = [_CTFrame new];
     if (range.length == 0) {
-        return lines;
+        return frame;
     }
 
     // Call custom renderer to get all glyph run details
+    ComPtr<IDWriteTextLayout> textLayout = __CreateDWriteTextLayout(ts, range, frameSize);
+    ComPtr<CustomDWriteTextRenderer> textRenderer = Make<CustomDWriteTextRenderer>();
     _DWriteGlyphRunDetails glyphRunDetails = {};
-    __GetGlyphRunDetails(ts, range, frameSize, &glyphRunDetails);
+    textLayout->Draw(&glyphRunDetails, textRenderer.Get(), 0, 0);
+    DWRITE_TEXT_METRICS textMetrics;
+    THROW_IF_FAILED(textLayout->GetMetrics(&textMetrics));
+    frame->_frameRect = frameSize;
+
+    // TODO:: find more precise value than 1.0 to increase width by to fully enclose frame
+    frame->_frameRect.size.width = std::min(frameSize.size.width, textMetrics.widthIncludingTrailingWhitespace + 1.0f);
```

Comment:
Dwrite won't write all of the text in the frame if it is within some small margin of the width required, but I'm not sure what the precise value is nor have I found any documentation on what it should be.

---
In reply to: [80596031](https://github.com/Microsoft/WinObjC/pull/1027#discussion_r80596031) [](ancestors = 80596031)